compiler/rustc_target/src/abi/call/x86.rs

```diff
@@ -53,8 +53,76 @@ where
-            arg.make_indirect_byval();
+
+        // FIXME: MSVC 2015+ will pass the first 3 vector arguments in [XYZ]MM0-2
+        // See https://reviews.llvm.org/D72114 for Clang behavior
+
+        let t = cx.target_spec();
+        let align_4 = Align::from_bytes(4).unwrap();
+        let align_16 = Align::from_bytes(16).unwrap();
+
+        if t.is_like_msvc
+            && arg.layout.is_adt()
+            && let Some(requested_align) = arg.layout.repr_options().align
+            && requested_align > align_4
+        {
+            // MSVC has special rules for overaligned arguments: https://reviews.llvm.org/D72114.
+            // Summarized here:
+            // - Arguments with _requested_ alignment > 4 are passed indirectly.
+            // - For backwards compatibility, arguments with natural alignment > 4 are still passed
+            //   on stack (via `byval`). For example, this includes `double`, `int64_t`,
+            //   and structs containing them, provided they lack an explicit alignment attribute.
+            assert!(arg.layout.align.abi >= requested_align,
+                "abi alignment {:?} less than requested alignment {:?}",
+                arg.layout.align.abi,
+                requested_align
+            );
+            arg.make_indirect();
+        } else if arg.layout.is_aggregate() {
+            // We need to compute the alignment of the `byval` argument. The rules can be found in
+            // `X86_32ABIInfo::getTypeStackAlignInBytes` in Clang's `TargetInfo.cpp`. Summarized
+            // here, they are:
+            //
+            // 1. If the natural alignment of the type is <= 4, the alignment is 4.
+            //
+            // 2. Otherwise, on Linux, the alignment of any vector type is the natural alignment.
+            // This doesn't matter here because we only pass aggregates via `byval`, not vectors.
+            //
+            // 3. Otherwise, on Apple platforms, the alignment of anything that contains a vector
+            // type is 16.
+            //
+            // 4. If none of these conditions are true, the alignment is 4.
+
+            fn contains_vector<'a, Ty, C>(cx: &C, layout: TyAndLayout<'a, Ty>) -> bool
+            where
+                Ty: TyAbiInterface<'a, C> + Copy,
+            {
+                match layout.abi {
+                    Abi::Uninhabited | Abi::Scalar(_) | Abi::ScalarPair(..) => false,
+                    Abi::Vector { .. } => true,
+                    Abi::Aggregate { .. } => {
+                        for i in 0..layout.fields.count() {
+                            if contains_vector(cx, layout.field(cx, i)) {
+                                return true;
+                            }
+                        }
+                        false
+                    }
+                }
+            }
+
+            let byval_align = if arg.layout.align.abi < align_4 {
+                // (1.)
+                align_4
```

Comment:
On further consideration, I'm a bit worried about the alignment mismatch with the source pointer here. LangRef says:
> The byval attribute also supports specifying an alignment with the align attribute. It indicates the alignment of the stack slot to form and the known alignment of the pointer specified to the call site. If the alignment is not specified, then the code generator makes a target-specific assumption.

If I'm understanding what is going to happen here correctly, then the pointer we're passing to the call may have alignment lower than 4, which would violate the "the known alignment of the pointer specified to the call site" part of this.

Clang does specify `align 4` on the byval argument, but it also uses `align 4` for the corresponding alloca: https://clang.godbolt.org/z/cW1cPrnsr However, Rust uses `align 4`  for the `alloca`: https://rust.godbolt.org/z/db7qor7es So I think we'll get a mismatch here.

Reply:
And this isn't working on accident, clang will also insert an explicit memcpy to a higher-aligned stack slot if necessary: https://clang.godbolt.org/z/5Yofxz3We

I believe this is handled by the code starting at https://github.com/llvm/llvm-project/blob/0e7cfa59ebe9d87029ebf75bb4a05316dd3c3d63/clang/lib/CodeGen/CGCall.cpp#L5060 (this is case 2).

I think we would need something like this as well.

Reply:
Yeah, we already have code to do this when passing packed struct fields. I'll extend it to this case as well.

I also plan to change `on_stack` to `Option<Align>` as eddyb originally suggested, since implicitly relying on `pointee_align` feels a bit more sketchy with this change.